Add cwd flag

kittens opened the pull request:
**Summary**

Fixes #692. This adds a `--cwd` flag to change the current working directory. This is so you can specify another working directory to perform an operation that's not the current one. But Sebastian, why not just cd into it and run Yarn there?

1. In Node the cwd is `realpath`d which means that if you cd into a symlinked directory then the cwd will be at the real location. I have a weird use-case where I have a symlinked folder in a workspace that is outside of the workspace root.
2. It makes Yarn wrapper scripts nicer since you don't need to mess with cding into a folder then cding back out. 

<!-- Is the feature a substantial feature request? Please use https://github.com/yarnpkg/rfcs -->

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

**Test plan**

Added tests.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

arcanis commented:
But Sebastian, why not just cd into it and run Yarn there? 😆 

Related to #692 - I'm personally not convinced this feature should be in Yarn itself, but if people really want it I see no point preventing this from landing altogether :) First point is interesting tho. Do you have a detailed example?

arcanis commented:
I see. It's unfortunate Node is realpathing the cwd 😞 The only thing that makes me cautious with this patch is that unless I'm mistaken, the yarnrc options will not be read relative to the option `cwd`, which might be quite unexpected. @BYK wdyt?

kittens commented:
> unless I'm mistaken, the yarnrc options will not be read relative to the option cwd

[`options.cwd` is passed down to all `Registry` instances](https://github.com/yarnpkg/yarn/blob/fa0fb69a567e1c44320d6f2a5f2a8ee42bc226ea/src/config.js#L245). Everything in the Yarn codebase references `config.cwd` instead of `process.cwd()` (at least it should).

arcanis commented:
Not the command line options. They are extract in [cli/index.js](https://github.com/yarnpkg/yarn/blob/3bfa1e3d6303c58aed9660e1a2af3a6d6f030bce/src/cli/index.js#L108) because we need to have them before we run the command line argument parser, which occurs before we create the configuration. Adding a way to change the cwd would in turn have a weird behavior, where half the options (all command line options) will be read from a file (the original directory yarnrc) whilst the others will be read from another file (the redirected dir yarnrc).

kittens commented:
Updated to handle `--cwd` inside of `.yarnrc`. This makes the initial args `.yarnrc` search work if you pass a `--cwd` on the CLI and when it finds a `--cwd` flag in a `.yarnrc` file then it will reset and load `.yarnrc` relative to it. It detects `.yarnrc` recursion too and will bail out. I've added a bunch of tests that cover the new and previous behaviour of the code.

BYK commented on src/util/rc.js:
```diff
@@ -55,13 +26,42 @@ export function findRc(name: string, parser: Function): Object {
+  while (true) {
+    configPaths.unshift(path.join(cwd, `.${name}rc`));
+
+    const upperCwd = path.dirname(cwd);
```
Shouldn't we stop at the project root?

arcanis replied:
Not really, because if we were to do that we wouldn't read the yarnrc files in $HOME, for example.

arcanis commented on src/rc.js:
```diff
@@ -25,57 +23,76 @@ const buildRcConf = () =>
+    }
   }
+  return null;
+}
```
Not completely sure about this one. Running `yarn run -- --cwd foo` will break, the `--cwd` will be interpreted when it really shouldn't. Additionally, making a pseudo-parsing here means that it might get out of sync if we change the cli parsing in some unexpected ways (for example #4152).

I feel like a better option might be to do a first pass of cli parsing using the yarnrc options available in the current directory branch, then extract the `--cwd` option and, if present, run a second pass using the new directory as base. It would always be synced, and wouldn't have much perf impact in most cases. I don't know of feasable this is with commander tho.

kittens replied:
It would require abstracting out a lot of stuff from `cli/index.js` which I'm not comfortable doing at the moment.

arcanis replied:
Makes sense. Can you at least iterate from start to end and stop iterating if you encounter a `--` token?

BYK commented on src/reporters/lang/en.js:
```diff
@@ -116,6 +116,8 @@ const messages = {
     'Pattern $0 is trying to unpack in the same destination $1 as pattern $2. This could result in a non deterministic behavior, skipping.',
   incorrectLockfileEntry: 'Lockfile has incorrect entry for $0. Ignoring it.',
+  cwdInYarnrcWarning:
+    'There was a --cwd flag in a .yarnrc file relative to this cwd. .yarnrc files relative to the --cwd have been ignored.',
```
Better to say "relative to where you ran yarn" instead of "relative to this cwd". What do you think?

arcanis replied:
I don't see this message being referenced anywhere - @kittens is it still used or can we remove it?

BYK commented on src/rc.js:
```diff
@@ -5,17 +5,15 @@ import parse from './lockfile/parse.js';
 
-    for (const key of keys) {
+    // some keys reference directories so keep their relativity
+    for (const key in values) {
```
Aren't `for ... in` loops "considered harmful? :D

kittens replied:
Not really. Mutated prototypes are the only dangerous thing and they aren't really a practical concern.

arcanis replied:
I'd say the problem is more readability than anything. It's really unusal to see `for ... in` loops nowadays, and it's not obvious when reading the code whether it's intended or not. Flow helps a bit, but not much.

That being said, `for ... in` loops are used here and there in the source code so if you want to use it there I'm fine with it. If we eventually choose to stop using them it will be in another PR.

arcanis commented on src/rc.js:
```diff
@@ -25,57 +23,76 @@ const buildRcConf = () =>
+  const newCwd = extractCwdArg(newArgs);
+  if (newCwd && newCwd !== origCwd) {
+    // ensure that we don't enter into a loop
+    if (previousCwds.includes(newCwd)) {
```
We need to use `indexOf` for the Node 4 compatibility

kittens replied:
I thought that too but `includes` is already used [here](https://github.com/yarnpkg/yarn/blob/master/src/lockfile/parse.js#L374).

arcanis replied:
It's because `String#includes` is implemented in Node 4, but not `Array#includes`

arcanis commented:
I pushed a change for the two things I think should be updated. Only remains the reporter key that doesn't seem used, so if you confirm me it's safe to remove it I'll do it and merge this PR if the tests pass 😃

taion commented:
Would it make sense to make `-C` be an alias for `--cwd`? npm's `--prefix` is not a great name for the flag, but if nothing else is in the way, having `-C` for compatibility seems like a free win.

Also, the `--cwd` flag appears to be missing from the online documentation.

BYK commented:
@taion a quick search in the code for `-C` revealed that this is used for `--caret` in `upgrade` and `upgrade-interactive` commands.  Although I agree that it would be nice to have a shortcut, it would be a breaking change at this point which I'm not willing to make.

BYK commented:
>Also, the --cwd flag appears to be missing from the online documentation.

Thanks for the heads up! Filed #651. Feel free to submit something yourself if you're feeling like it :)

guilhermecvm commented:
Doesn't work with symlinks (trying to run from root package.json when using yarn workspaces)
```json
{
  "scripts": {
    "start": "yarn --cwd packages/dolphino react-scripts start"
  }
}
```
```
➜ dolphino (master) $ yarn start
yarn run v1.3.2
$ yarn --cwd packages/dolphino react-scripts start
$ packages/dolphino/node_modules/.bin/react-scripts start
/bin/sh: packages/dolphino/node_modules/.bin/react-scripts: No such file or directory
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
````

But file do exists and is symlinked
```
➜ dolphino (master) $ ll packages/dolphino/node_modules/.bin/react-scripts
lrwxr-xr-x  1 guilhermecvm  staff    59B 15 Nov 20:02 packages/dolphino/node_modules/.bin/react-scripts -> ../../../../node_modules/react-scripts/bin/react-scripts.js
```

BYK commented:
@guilhermecvm please file a new bug report so this can be tracked and addressed.

arcanis commented:
Or better yet, a pull request 👐

AlastairTaft commented:
I hit a similar issue where I've symlinked to another mono-repo to include all the packages there as dependencies but that tries to run babel as part of the install, unfortunately babel uses `process.cwd()` so it falls over. Can't work out if this will solve my use case.
